misc/pacstall.8
.\" Automatically generated by Pandoc 3.1.3
.\"
.\" Define V font for inline verbatim, using C font in formats
.\" that render this, and otherwise B font.
.ie "\f[CB]x\f[]"x" \{\
. ftr V B
. ftr VI BI
. ftr VB B
. ftr VBI BI
.\}
.el \{\
. ftr V CR
. ftr VI CI
. ftr VB CB
. ftr VBI CBI
.\}
.TH "Pacstall" "8" "July 8, 2024" "Pacstall 5.3.2 Verdigris" "General Commands Manual"
.hy
.SH NAME
.PP
Pacstall - The AUR for Ubuntu
.SH SYNOPSIS
.PP
\f[B]pacstall -V\f[R]
.PP
\f[B]pacstall\f[R] [\f[B]-P -K -B -Q -Nc -Ns\f[R]] \f[B]-I\f[R]
\f[I]package\f[R] \&...
.PP
\f[B]pacstall -S\f[R] \f[I]term\f[R]
.PP
\f[B]pacstall\f[R] [\f[B]-P\f[R]] \f[B]-R\f[R] \f[I]package\f[R] \&...
.PP
\f[B]pacstall -D\f[R] \f[I]package\f[R] \&...
.PP
\f[B]pacstall -A\f[R] \f[I]repository\f[R]
.PP
\f[B]pacstall -Rr\f[R] \f[I]repository\f[R]
.PP
\f[B]pacstall -U\f[R] [\f[I]username branch\f[R]]
.PP
\f[B]pacstall -L\f[R]
.PP
\f[B]pacstall\f[R] [\f[B]-P -K\f[R]] \f[B]-Up\f[R]
.PP
\f[B]pacstall -Qi\f[R] \f[I]package\f[R] [\f[I]key\f[R]]
.PP
\f[B]pacstall -Qa\f[R]
\f[I]package\f[R]#\f[I]number\f[R](\[at]*metalink*)
.PP
\f[B]pacstall -T\f[R] \f[I]package\f[R]
.SH DESCRIPTION
.PP
Pacstall is the AUR Ubuntu wishes it had.
It takes the AUR and puts a spin on it, making it easier to install
programs without scouring github repositories and the likes.
.SH OPTIONS
.TP
\f[B]-h\f[R], \f[B]--help\f[R]
Show the help message and exit.
.TP
\f[B]-I\f[R], \f[B]--install\f[R]
Installs a given package from a repository, or a pacscript.
The \f[B]-I\f[R] flag has the ability to install packages from an
arbitrary repository, defaulting to GitHub.
If multiple repositories are enabled, \f[B]pacstall\f[R] will default to
\f[B]pacstall/pacstall-programs\f[R], but you can still choose to
install from another repository.
You can also specify a local pacscript by including the `.pacscript'
suffix.
The following are valid examples of the \f[B]-I\f[R] flag:
.RS
.RS
.PP
$ \f[B]pacstall -I neofetch\f[R] # Install \f[B]neofetch\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -I\f[R]
\[lq]\f[B]alacritty\f[R]\[at]\f[I]user\f[R]/\f[I]reponame\f[R]\[rq] #
Install \f[B]alacritty\f[R] from
<\f[B]https://github.com/user/reponame\f[R]>
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -I neofetch.pacscript\f[R] # Install \f[B]neofetch\f[R]
from local pacscript
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -I ${parent}:pkgbase\f[R] # Install all pkgnames in the
pkgbase named \f[B]parent\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -I
\f[BI]p\f[B]\f[BI]a\f[B]\f[BI]r\f[B]\f[BI]e\f[B]\f[BI]n\f[B]\f[BI]t\f[B]:{child}\f[R]
# Install only the pkgname \f[B]child\f[R] from the pkgbase
\f[B]parent\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -I package.pacscript:${child}\f[R] # Install only the
pkgname \f[B]child\f[R] from a local split package
.RE
.RE
.TP
\f[B]-R\f[R], \f[B]--remove\f[R]
Removes a given package from system.
For example:
.RS
.RS
.PP
$ \f[B]pacstall -R\f[R] \f[I]neofetch\f[R]
.RE
.RE
.TP
\f[B]-S\f[R], \f[B]--search\f[R]
Searches for a given package from all available repositories.
A clickable hyperlink to the repositories containing the packages are
displayed in VTE terminals.
Other repositories can be added using the \f[B]-A\f[R] flag.
.RS
.RS
.PP
$ \f[B]pacstall -S\f[R] \f[I]-git\f[R] # search for all git packages
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -S\f[R] \f[I]-bin\f[R] # search for all binary packages
.RE
.RE
.TP
\f[B]-A\f[R], \f[B]--add-repo\f[R]
Add a given repository to \f[B]pacstall\f[R].
If a plain GitHub or GitLab link is given without a \f[I]branch\f[R],
\f[B]pacstall\f[R] will default to \f[B]master\f[R], but you can specify
a \f[I]branch\f[R].
Local repositories are also supported, which can be useful for testing
pacscripts without pushing remotely, having a portable repository on a
USB stick, custom/private applications, etc.
The following commands are valid ways to add a repository:
.RS
.RS
.PP
$ \f[B]pacstall -A https://github.com\f[R]/\f[I]user\f[R]/\f[I]repo\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A
https://github.com\f[R]/\f[I]user\f[R]/\f[I]repo\f[R]/\f[B]tree\f[R]/\f[I]branch\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A github:user/repo\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A github:user/repo#branch\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A https://gitlab.com\f[R]/\f[I]user\f[R]/\f[I]repo\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A
https://gitlab.com\f[R]/\f[I]user\f[R]/\f[I]repo\f[R]/\f[B]-\f[R]/\f[B]tree\f[R]/\f[I]branch\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A
https://raw.githubusercontent.com\f[R]/\f[I]user\f[R]/\f[I]repo\f[R]/\f[I]master\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A
https://gitlab.com\f[R]/\f[I]user\f[R]/\f[I]repo\f[R]/\f[B]-\f[R]/\f[B]raw\f[R]/\f[I]master\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A gitlab:user/repo\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A gitlab:user/repo#branch\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A file:///home/user/local-repository\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A /home/user/local-repository\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A local:/home/user/local-repository\f[R]
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -A
https://myonlinerepository.com/directory_containing_packagelist\f[R]
.RE
.RE
.TP
\f[B]-Rr\f[R], \f[B]--remove-repo\f[R]
Remove a given repository from \f[B]pacstall\f[R].
See \f[I]-A\f[R], \f[I]--add-repo\f[R] for syntax usage, swapping
\f[I]-A\f[R] with \f[I]-Rr\f[R].
.TP
\f[B]-U\f[R], \f[B]--update\f[R]
Updates \f[B]pacstall\f[R] and needed scripts.
Arguments to this flag are as follows: The \f[I]first\f[R] flag is the
\f[I]username\f[R], while the \f[I]second\f[R] flag is the
\f[I]branch\f[R].
The second argument is optional as long as the first argument is given,
however both the first and second arguments are optional together.
If no argument for the \f[I]branch\f[R] is given, it will default to
\f[B]master\f[R].
When updating to a specific repository, the arguments used will be saved
so that the next time the \f[B]-U\f[R] flag is used, if no arguments are
given, \f[B]pacstall\f[R] will update through the same user and branch.
If there is a typo in \f[I]username\f[R] or \f[I]branch\f[R],
\f[B]pacstall\f[R] will not update, and your current state will not be
touched.
You may also replace the \f[I]username\f[R] and \f[I]branch\f[R]
arguments with a \f[I].\f[R], which will update from a local Pacstall
repository.
The following are valid examples:
.RS
.RS
.PP
$ \f[B]pacstall -U\f[R] \f[I]pacstall\f[R] \f[I]develop\f[R] # This will
update pacstall from
<\f[B]https://github.com/pacstall/pacstall/tree/develop\f[R]>
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -U\f[R] \f[I]pacstall\f[R]:\f[I]develop\f[R] # This will
update pacstall from
<\f[B]https://github.com/pacstall/pacstall/tree/develop\f[R]>
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -U\f[R] # This will update pacstall from
<\f[B]https://github.com/pacstall/pacstall/tree/develop\f[R]>, because
the previous arguments were saved
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -U\f[R] \f[I]user\f[R] # This will update pacstall from
<\f[B]https://github.com/user/pacstall/tree/master\f[R]>
.RE
.RE
.RS
.RS
.PP
$ \f[B]pacstall -U .\f[R] # This will update pacstall from a local
Pacstall repository
.RE
.RE
.TP
\f[B]-Up\f[R], \f[B]--upgrade\f[R]
Upgrade packages that have a newer version.
.TP
\f[B]-L\f[R], \f[B]--list\f[R]
List installed packages.
.TP
\f[B]-T\f[R], \f[B]--tree\f[R]
Display a tree graph of a package.
.TP
\f[B]-V\f[R], \f[B]--version\f[R]
Lists \f[B]pacstall\f[R] version and name.
.TP
\f[B]-D\f[R], \f[B]--download\f[R]
Download pacscript to current directory.
You can specify an arbitrary repository like so:
.RS
.RS
.PP
$ \f[B]pacstall -D neofetch\f[R]\[at]\f[I]user\f[R]/\f[I]reponame\f[R] #
This downloads the neofetch pacscript from
<\f[B]https://github.com/user/reponame\f[R]>
.RE
.RE
.TP
\f[B]-Qi\f[R], \f[B]--query-info\f[R]
Display metadata of an installed package.
Supply a key from the output to get it\[cq]s value.
If a key has spaces in it, replace them with an underscore.
For example:
.RS
.RS
.PP
$ \f[B]pacstall -Qi\f[R] \f[I]neofetch\f[R] \f[I]install_type\f[R]
.RE
.RE
.TP
\f[B]-Qa\f[R], \f[B]--quality-assurance\f[R]
Test a package from a PR downstream, before it is merged.
Used like:
.RS
.RS
.PP
$ \f[B]pacstall -Qa\f[R] \f[I]firefox-bin\f[R]#\f[I]5853\f[R] $
\f[B]pacstall -Qa\f[R]
\f[I]firefox-bin\f[R]#\f[I]5853\f[R]\[at]\f[I]github:pacstall/pacstall-programs\f[R]
$ \f[B]pacstall -Qa\f[R]
\f[I]firefox-bin\f[R]\[at]\f[I]github:pacstall/pacstall-programs\f[R]#\f[I]5853\f[R]
.RE
.RE
.TP
: Where the \f[I]package\f[R] is given first, followed by the pull request \f[I]NUM\f[R], separated by a \f[I]#\f[R].
Optionally, a \f[I]metalink\f[R] separated by a \f[I]\[at]* may be
provided before or after the \f[R]#NUM*.
The \f[I]metalink\f[R] is broken down into 3 parts: \f[I]provider\f[R],
\f[I]owner\f[R], and \f[I]repo\f[R].
The \f[I]owner\f[R] should be the owner of the repository the PR is
being merged into, \f[I]not\f[R] the user who created the PR.
No other flags should be provided to this command.
.TP
\f[B]-P\f[R], \f[B]--disable-prompts\f[R]
Add this flag alongside other commands to disable prompts and accept all
defaults.
For example:
.RS
.RS
.PP
$ \f[B]pacstall -PR\f[R] \f[I]neofetch\f[R]
.RE
.RE
.TP
\f[B]-K\f[R], \f[B]--keep\f[R]
Add this flag alongside \f[B]-I\f[R] and \f[B]-Up\f[R] to keep the build
of a package if the build process fails or succeeds.
.TP
\f[B]-B\f[R], \f[B]--build-only\f[R]
Add this flag alongside \f[B]-I\f[R] and \f[B]-Up\f[R] to just build the
deb, and not install.
.TP
\f[B]-Q\f[R], \f[B]--quiet\f[R]
Add this flag alongside \f[B]-I\f[R] and \f[B]-Up\f[R] to silence
downloading.
.TP
\f[B]-Nc\f[R], \f[B]--nocheck\f[R]
Add this flag alongside \f[B]-I\f[R] and \f[B]-Up\f[R] to skip the
check() function when building.
.TP
\f[B]-Ns\f[R], \f[B]--nosandbox\f[R]
Add this flag alongside \f[B]-I\f[R] and \f[B]-Up\f[R] to build without
bwrap.
This flag should be used with caution, and can lead to potential
unwanted harm on a system.
It is intended for use in environments such as chroot which have other
levels of isolation.
.SH ENVIRONMENT
.TP
\f[B]DISABLE_PROMPTS\f[R]
Equivalent of the \f[B]-P\f[R] flag.
.TP
\f[B]PACSTALL_SUPPRESS_SOLUTIONS\f[R]
Set this to remove suggestions given by Pacstall to fix problems.
.TP
\f[B]PACSTALL_BUILD_CORES\f[R]
Set this to an integer to override the \f[V]nproc\f[R] command and
\f[V]${NCPU}\f[R] variable in any Pacscript.
.TP
\f[B]PACSTALL_EDITOR\f[R]
Sets the editor used by pacstall.
.TP
\f[B]PACSTALL_DOWNLOADER\f[R]
Sets the downloading tool used.
Can be set to \f[V]axel\f[R], \f[V]wget\f[R], \f[V]quiet-wget\f[R], or
\f[V]curl\f[R].
.TP
\f[B]PACSTALL_PAYLOAD\f[R]
Sets the path to a pre-downloaded package archive.
.TP
\f[B]NO_COLOR\f[R]
Turns off all colors.
.SH FILES
.TP
\f[B]/usr/share/pacstall/repo/pacstallrepo\f[R]
Repository database.
.TP
\f[B]/usr/share/pacstall/repo/update\f[R]
Saved arguments used by the \f[B]-U\f[R] flag.
.TP
\f[B]/usr/share/pacstall/scripts\f[R]
Assorted scripts used by \f[B]pacstall\f[R] to operate.
.SH BUGS
.PP
If you find any bugs in Pacstall itself, please submit a bug request at
<\f[B]https://github.com/pacstall/pacstall/issues\f[R]>.
.PP
If you find any bugs in a Pacscript, please submit a bug request at
<\f[B]https://github.com/pacstall/pacstall-programs/issues\f[R]>.
.PP
The main method Pacstall developers and users uses to communicate with
each other is through Discord, but you may also communicate through
GitHub issues.
.SH AUTHORS
.PP
The Pacstall team (pacstall\[at]pm.me)
.IP \[bu] 2
Elsie19
.IP \[bu] 2
wizard-28
.IP \[bu] 2
D-Brox
.IP \[bu] 2
saenai255
.IP \[bu] 2
oklopfer
.IP \[bu] 2
0oAstro


pacstall/pacstall.py
#!/bin/env python3

"""Main Pacstall launcher file."""

#     ____                  __        ____
#    / __ \____ ___________/ /_____ _/ / /
#   / /_/ / __ `/ ___/ ___/ __/ __ `/ / /
#  / ____/ /_/ / /__(__  ) /_/ /_/ / / /
# /_/    \__,_/\___/____/\__/\__,_/_/_/
#
# Copyright (C) 2020-2021
#
# This file is part of Pacstall
#
# Pacstall is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, version 3 of the License
#
# Pacstall is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License

import sys
from argparse import ArgumentParser, HelpFormatter, Namespace
from fcntl import LOCK_EX, LOCK_NB, lockf
from getpass import getuser
from subprocess import call
from time import sleep

from rich.traceback import install

from api import message
from api.color import Foreground
from cmds import download


# Copied from https://stackoverflow.com/a/23941599 and modified
class CustomHelpFormatter(HelpFormatter):
    """
    Custom help message formatter for Pacstall.

    Format:
    -s, --long       help message
    """

    def _format_action_invocation(self, action) -> str:  # type: ignore[no-untyped-def]
        if not action.option_strings:
            (metavar,) = self._metavar_formatter(action, action.dest)(1)
            return metavar

        parts = []
        # if the Optional doesn't take a value, format is:
        #    -s, --long
        if action.nargs == 0:
            parts.extend(action.option_strings)

        # if the Optional takes a value, format is:
        #    -s ARGS, --long ARGS
        # change to
        #    -s, --long
        else:
            for option_string in action.option_strings:
                parts.append("%s" % option_string)
        return ", ".join(parts)


def parse_arguments() -> Namespace:
    """
    Parses command line arguments passed to Pacstall.

    Prints help and exits if no argument is passed.

    Returns
    -------
    Namespace: Containing all the parsed arguments
    """
    parser = ArgumentParser(prog="pacstall", formatter_class=CustomHelpFormatter)
    commands = parser.add_argument_group("commands").add_mutually_exclusive_group()
    modifiers = parser.add_argument_group("modifiers")

    commands.add_argument(
        "-I", "--install", metavar="package", nargs="+", help="install packages"
    )
    commands.add_argument(
        "-S", "--search", metavar="package", nargs="?", help="search for packages"
    )
    commands.add_argument(
        "-R", "--remove", metavar="package", nargs="?", help="remove packages"
    )
    commands.add_argument(
        "-D", "--download", metavar="package", nargs="?", help="download pacscripts"
    )
    commands.add_argument(
        "-A", "--add-repo", metavar="repo", nargs="?", help="add repos to the repo list"
    )
    commands.add_argument(
        "-V",
        "--version",
        action="version",
        version=f"{Foreground.BIBLUE}Pacstall {Foreground.BIWHITE}2.0 {Foreground.BIYELLOW}Kournikova",
        help="show version",
    )
    commands.add_argument(
        "-L", "--list", action="store_true", help="list installed packages"
    )
    commands.add_argument(
        "-Up", "--upgrade", action="store_true", help="upgrade packages"
    )
    commands.add_argument(
        "-Qi", "--query-info", metavar="package", nargs=1, help="show package info"
    )

    modifiers.add_argument(
        "-P",
        "--disable-prompts",
        dest="disable_prompts",
        action="store_true",
        help="disable prompts for unattended operations",
    )
    modifiers.add_argument(
        "-K",
        "--keep",
        dest="keep",
        action="store_true",
        help="retain build directory after installation",
    )

    if len(sys.argv) == 1:
        parser.print_help()
        sys.exit(1)

    return parser.parse_args()


if __name__ == "__main__":
    install(
        show_locals=True
    )  # --> Install Rich's traceback handler for better looking tracebackes

    if getuser() == "root":
        message.fancy("error", "Pacstall cannot be run as root")
        sys.exit(1)

    args = parse_arguments()

    if args.install or args.remove or args.upgrade:
        lock_file = open("/var/lock/pacstall.lock", "w")
        call(["/usr/bin/sudo", "/usr/bin/chown", "root", "/var/lock/pacstall.lock"])
        while True:
            try:
                lockf(lock_file, LOCK_EX | LOCK_NB)
                break
            except OSError:
                message.fancy("error", "Pacstall is already running another instance")
                sleep(1)

    if args.download:
        sys.exit(download.execute(args.download))
